app/src/behaviors/behavior_smart_interrupt.c

```diff
@@ -0,0 +1,288 @@
+    const struct behavior_smart_interrupt_config *cfg = dev->config;
+    struct active_smart_interrupt *smart_interrupt;
+    smart_interrupt = find_smart_interrupt(event.position);
+    if (smart_interrupt == NULL) {
+        if (new_smart_interrupt(event.position, cfg, &smart_interrupt) == -ENOMEM) {
+            LOG_ERR("Unable to create new smart_interrupt. Insufficient space in "
+                    "active_smart_interrupts[].");
+            return ZMK_BEHAVIOR_OPAQUE;
+        }
+        LOG_DBG("%d created new smart_interrupt", event.position);
+    }
+    LOG_DBG("%d smart_interrupt pressed", event.position);
+    stop_timer(smart_interrupt);
+    smart_interrupt->is_pressed = true;
+    if (smart_interrupt->first_press) {
+        behavior_keymap_binding_pressed(&cfg->behaviors[0], event);
+        behavior_keymap_binding_released(&cfg->behaviors[0], event);
+        smart_interrupt->first_press = false;
+    }
+    behavior_keymap_binding_pressed(&cfg->behaviors[1], event);
+    return ZMK_BEHAVIOR_OPAQUE;
+}
+
+static int on_smart_interrupt_binding_released(struct zmk_behavior_binding *binding,
+                                               struct zmk_behavior_binding_event event) {
+    const struct device *dev = device_get_binding(binding->behavior_dev);
+    const struct behavior_smart_interrupt_config *cfg = dev->config;
+    LOG_DBG("%d smart_interrupt keybind released", event.position);
+    struct active_smart_interrupt *smart_interrupt = find_smart_interrupt(event.position);
+    if (smart_interrupt == NULL) {
+        return ZMK_BEHAVIOR_OPAQUE;
+    }
+    smart_interrupt->is_pressed = false;
+    behavior_keymap_binding_released(&cfg->behaviors[1], event);
+    reset_timer(k_uptime_get(), smart_interrupt);
+    return ZMK_BEHAVIOR_OPAQUE;
+}
+
+static int behavior_smart_interrupt_init(const struct device *dev) {
+    static bool init_first_run = true;
+    if (init_first_run) {
+        for (int i = 0; i < ZMK_BHV_MAX_ACTIVE_SMART_INTERRUPTS; i++) {
+            k_work_init_delayable(&active_smart_interrupts[i].release_timer,
+                                  behavior_smart_interrupt_timer_handler);
+            clear_smart_interrupt(&active_smart_interrupts[i]);
+        }
+    }
+    init_first_run = false;
+    return 0;
+}
+
+static const struct behavior_driver_api behavior_smart_interrupt_driver_api = {
+    .binding_pressed = on_smart_interrupt_binding_pressed,
+    .binding_released = on_smart_interrupt_binding_released,
+};
+
+static int smart_interrupt_position_state_changed_listener(const zmk_event_t *eh);
+static int smart_interrupt_layer_state_changed_listener(const zmk_event_t *eh);
+
+ZMK_LISTENER(behavior_smart_interrupt, smart_interrupt_position_state_changed_listener);
+ZMK_SUBSCRIPTION(behavior_smart_interrupt, zmk_position_state_changed);
+
+ZMK_LISTENER(behavior_smart_interrupt2, smart_interrupt_layer_state_changed_listener);
+ZMK_SUBSCRIPTION(behavior_smart_interrupt2, zmk_layer_state_changed);
```

Comment:
Why two different listeners? Can have one listener subscribed to multiple events.